src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsModelValidator.cs

```diff
@@ -94,31 +94,52 @@ public IEnumerable<ModelValidationResult> Validate(ModelValidationContext valida
-                // (person.Name.Name). However the invoking validator does not have a way to distinguish between these
-                // two cases. Consequently we'll only set MemberName if this validation returns a MemberName that is
-                // different from the property being validated.
-
-                var errorMemberName = result.MemberNames.FirstOrDefault();
-                if (string.Equals(errorMemberName, memberName, StringComparison.Ordinal))
-                {
-                    errorMemberName = null;
-                }
-
-                string errorMessage = null;
+                string errorMessage;
                 if (_stringLocalizer != null &&
                     !string.IsNullOrEmpty(Attribute.ErrorMessage) &&
                     string.IsNullOrEmpty(Attribute.ErrorMessageResourceName) &&
                     Attribute.ErrorMessageResourceType == null)
                 {
-                    errorMessage = GetErrorMessage(validationContext);
+                    errorMessage = GetErrorMessage(validationContext) ?? result.ErrorMessage;
+                }
+                else
+                {
+                    errorMessage = result.ErrorMessage;
+                }
+
+                var validationResults = new List<ModelValidationResult>();
+                if (result.MemberNames != null)
+                {
+                    foreach (var resultMemberName in result.MemberNames)
+                    {
+                        // ModelValidationResult.MemberName is used by invoking validators (such as ModelValidator) to
+                        // append construct the ModelKey for ModelStateDictionary. When validating at type level we
+                        // want the returned MemberNames if specified (e.g. "person.Address.FirstName"). For property
+                        // validation, the ModelKey can be constructed using the ModelMetadata and we should ignore
+                        // MemberName (we don't want "person.Name.Name"). However the invoking validator does not have
+                        // a way to distinguish between these two cases. Consequently we'll only set MemberName if this
+                        // validation returns a MemberName that is different from the property being validated.
+                        ModelValidationResult validationResult;
+                        if (string.Equals(resultMemberName, memberName, StringComparison.Ordinal))
+                        {
+                            validationResult = new ModelValidationResult(memberName: null, message: errorMessage);
+                        }
+                        else
+                        {
+                            validationResult = new ModelValidationResult(resultMemberName, errorMessage);
+                        }
+
+                        validationResults.Add(validationResult);
```

Comment:
nit: This whole thing could probably just be decomposed into a ternary if you want it to be less verbose.